### ISSUE TITLE
Ill-typed tactics running

### ISSUE THREAD

**Issue by cpitclaudel** [FStarLang/FStar#1285]
Repro:

```
module PatternMatching

open FStar.Tactics

let example (p1 p2: Type0) =
  assert_by_tactic (p1 ==> (p1 ==> p2) ==> p2)
                   (_ <-- implies_intros;
                    match [1] with
                    | [] -> idtac)
```

**Comment by cpitclaudel** [comment]
Also causes "not an embedded string" errors when a string is built using an incomplete pattern-matching

**Comment by mtzguido** [comment]
Making a note.

To fix this:
1) don't use failwiths for unembedding, and allow them to return `None` (and get stuck).
2) check why we're not aborting due to tactic ill-typing, again.

**Comment by mtzguido** [comment]
OK, this will now not run due to ill-typing, and I'm more convinced that nothing will (unless of course we're in lax mode or admitting queries). And if it did run, you would get the following:
```
/home/guido/r/fstar/ulib/FStar.Tactics.Effect.fst(19,13-25,47): (Warning) Not an embedded tactic result: (match
    reify ((let Prims.Nil #uu___682578 = Prims.Cons 1 Prims.Nil in FStar.Tactics.Derived.idtac) ())
      "(alien:tactics.embed_proofstate)"
  with
  | FStar.Tactics.Result.Success #uu___572402 (#uu___572398, #uu___572399, #a, #q) ->
    (let () = FStar.Tactics.Types.tracepoint q in
      reify (FStar.Tactics.Effect.return a ()) (FStar.Tactics.Types.decr_depth q))
    <:
    FStar.Tactics.Result.__result Prims.unit
  | FStar.Tactics.Result.Failed #uu___573030 (#uu___573026, #uu___573027, #msg, #q) ->
    FStar.Tactics.Result.Failed (FStar.Pervasives.Native.Mktuple2 msg q))
<:
FStar.Tactics.Result.__result Prims.unit
./P.fst(5,0-10,34): (Error) Tactic got stuck! Please file a bug report with a minimal reproduction of this issue.
(match
    reify ((let Prims.Nil #uu___682578 = Prims.Cons 1 Prims.Nil in FStar.Tactics.Derived.idtac) ())
      "(alien:tactics.embed_proofstate)"
  with
  | FStar.Tactics.Result.Success #uu___572402 (#uu___572398, #uu___572399, #a, #q) ->
    (let () = FStar.Tactics.Types.tracepoint q in
      reify (FStar.Tactics.Effect.return a ()) (FStar.Tactics.Types.decr_depth q))
    <:
    FStar.Tactics.Result.__result Prims.unit
  | FStar.Tactics.Result.Failed #uu___573030 (#uu___573026, #uu___573027, #msg, #q) ->
    FStar.Tactics.Result.Failed (FStar.Pervasives.Native.Mktuple2 msg q))
<:
FStar.Tactics.Result.__result Prims.unit
1 error was reported (see above)
```
In this case the warning and the error exactly coincide, so it looks redundant, but it need not be so given that we may fail to unembed only subterms.

Going to master when CI allows.
(commits 77083db027c02d12ab9712acf43a9d172cccf239 and ab548e5f575fb8f16243e45b52e1feb4750abff6)

**Comment by cpitclaudel** [comment]
Thanks! So now the pattern match will be rejected, right?

I've been thinking about this a little, and I think we should distinguish between two kind of errors in tactics: hard errors, which stop the tactic execution, and tactic failures, which can be captured by trytac.  An incorrect pattern matching could raise a hard error, one that's not catch-able.

In Coq, error catching in tactics is a major source of confusion. Coq's tactic engine backtracks across all sort of errors, including references to undefined identifiers. It would be nice to not have the same shortcomings :)

**Comment by mtzguido** [comment]
There's two available contexts, the original one where the tactic started running, and the goal one. But the goal could possibly not exist, and it'd be nice to support normalizing outside of a goal. I'm thinking of having
```
norm_term_env : env -> term -> tactic unit // builtin

norm_term : term -> tactic unit = fun t ->
  e <-- cur_env;
  norm_term_env e t
```

**Comment by cpitclaudel** [comment]
Right, that makes sense. But how do you access the environment in which the tactic started running ?

**Comment by mtzguido** [comment]
It's in the proofstate, not accesible from userland currently, but easy enough to do.

**Comment by mtzguido** [comment]
Ugh, I mixed up the issues.. I meant to comment on #1299.

**Comment by mtzguido** [comment]
Maybe I should actually answer to your previous comment :).

In our case, a failed pattern match gets the tactic stuck, and that's not a recoverable failure. Recoverable failures are tactic computations that run just OK and end up with `Failed(msg, ps)`, and can thus only be introduced by `fail` and some builtins.

So failed matches are a hard error, but a user cannot *introduce* a hard error. Maybe we want an extra constructor `Aborted` for `result`, so a user can signal a non-recoverable failure. (Unsure if any builtin should hard fail: one can always make a hard variant).

WDYT?

**Comment by cpitclaudel** [comment]
That sounds perfect. I don't think users should be able to introduce unrecoverable errors.
But it'd be nice to have a finer-grained system, somehow.  Like in C#/Java/... you can catch families of exceptions. Coq has a related notion of failing more than one level. We could have a notion of errors that backtrack and errors that don't backtrack (I'd like the tactic monad to raise proper exceptions rather than strings)

**Comment by cpitclaudel** [comment]
Right. What I meant to emphasize is that errors are nice for control flow sometimes, and in Coq's model they get swallowed by the backtracking.

**Comment by mtzguido** [comment]
I consider this fixed, so closing and moving to #1304 for the error handling.

**Comment by mtzguido** [comment]
Sadly, I need to disable tactic re-typechecking (to unblock like 5 other issues) which causes this to regress. Reopening, and I'll try to find a solution that doesn't involve re-typechecking as time permits.

**Comment by mtzguido** [comment]
I restored this after fixing some other bugs.
